Allow system containers to have limits

vmarmol opened the pull request:
Not quite ready for review, the ContainerManager needs some work to actually use the new generalized ExternalContainer.

This PR will not set any limits, it will only provide the means to reserve resources for node services.

/cc @rjnagal  @dchen1107 

Will ping when more ready for review :)

vmarmol commented:
Ping @rjnagal @dchen1107 this PR is now ready for review.

As it currently is, it is a refactoring that has no functional change. It does, however, set the groundwork for allowing external containers to reserve parts of the machine.

Ran e2e and it came back green!

vmarmol commented:
Depends on #8681 btw, the first 3 commits are from that PR.

rjnagal commented:
Added comments to the new patches.

vmarmol commented:
All comments addressed PTAL @rjnagal and thanks for the review!

vmarmol commented:
Rebased and pushed.

vmarmol commented:
After our offline discussion I removed the last commit (no taking away from the capacity) and removed memory. PTAL @dchen1107 @rjnagal :)

dchen1107 commented:
#8681 gives us fully introspection stats on a given node, so that the user can understand the node better when the node is really overcommit through the monitoring pipeline. In a long run, those stats can help with capacity planning too. This one tries to apply some reservation to node services (not kubernetes daemons, not pods managed by kubernetes), but I have a real concerns real concerns, especially including it to 1.0 release:

1) The initial goal of this one is preventing us from overcommit and keep node stability. But I double this will achieve such goal. Resource limit is not mandatory in 1.0 release, most of users might not provide such information, and we don't have automated system to help them neither. In this case, applying some reservation to node services doesn't help us at all.
2) Another concern I had is that this PR subtracts reservation of /system from a node capacity at node level, and propagate such information to higher level. I think this will eventually make capacity planning harder. We had some experience with our internal system on this. I think with OSS, this will be even harder to manage for cluster admins.
3) Since capacity might change based on per node configuration introduced here, if the user does apply limit to their pods, they might wonder why my pods is not scheduled. Hiding such information at kubelet level makes explanation of why much harder. I would rather we have a policy for /system ready and controlled at cluster level, and pushed down to node, so that scheduler will respect /system reservation  against stable node capacity. 
4) ...

Edited after talking with @jnagal @vmarmol:

We agreed there is a value of this PR. I introduced LimitRange to cluster by default, which applies default cpu shares to all pods. But this is no way to apply configurable cpu.shares to /system based on node capacity. By default, /system will get 1024 value from kernel. The agreement we had:

1) Remove capacity change
2) Remove memory reservation from /system avoid accidentally applying to memory.

rjnagal commented on pkg/kubelet/container_manager_linux.go:
```diff
@@ -34,73 +36,132 @@ import (
+
+	// Sum up resources of all external containers.
+	for _, cont := range cm.systemContainers {
+		cpuReservation += cont.cpuMillicores
```
right now, we don't set cpuMillicores. Is it going to return 0 or 1000 for every container?

vmarmol replied:
0 since it is a Go-zeroed struct.

rjnagal commented:
LGTM, just one q about the reservation (should be called limit :) ) calculation.
Will let Dawn put her LGTM since she has been looking at it too.

dchen1107 commented:
LGTM

vmarmol committed:
Introduce SystemContainer to Kubelet ContainerManager.

This generalizes the handling of containers in the
ContainerManager.

Also introduces the ability to determine how much
resources are reserved for those system containers.

vmarmol commented:
Renamed to limit :)

erictune commented:
@vmarmo am I right in thinking this is post-1.0

vmarmol commented:
This PR has two features originally:
- Allow system containers (e.g.: Kubelet, Docker-daemon, System) to have limits
- Remove system container limits from capacity

The second is post v1.0 and was removed from the PR. The first is still relevant and for v1.0

vmarmol commented:
@dchen1107 @rjnagal @erictune I think this PR is ready to be merged, let me know if that is not the case :)

dchen1107 commented:
I agree. @vmarmol could you please update the pr title to reflect the real change in it? Thanks!
